spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/BoundStatementBuilder.scala

```diff
@@ -0,0 +1,51 @@
+package com.datastax.spark.connector.writer
+
+import com.datastax.driver.core._
+import com.datastax.spark.connector.types.ColumnType
+import com.datastax.spark.connector.util.Logging
+
+private[connector] class BoundStatementBuilder[T](
+    val rowWriter: RowWriter[T],
+    val preparedStmt: PreparedStatement,
+    val protocolVersion: ProtocolVersion) extends Logging {
+
+  private val columnNames = rowWriter.columnNames.toIndexedSeq
+  private val columnTypes = columnNames.map(preparedStmt.getVariables.getType)
+  private val converters = columnTypes.map(ColumnType.converterToCassandra(_)(protocolVersion))
+  private val buffer = Array.ofDim[Any](columnNames.size)
+
+  /** Creates `BoundStatement` from the given data item */
+  def bind(row: T): RichBoundStatement = {
+    val boundStatement = new RichBoundStatement(preparedStmt)
+    rowWriter.readColumnValues(row, buffer)
+    var bytesCount = 0
+    for (i <- 0 until columnNames.size) {
+      val converter = converters(i)
+      val columnName = columnNames(i)
+      val columnValue = converter.convert(buffer(i))
+      val columnType = columnTypes(i)
+      val serializedValue =
+        if (columnValue != null) columnType.serialize(columnValue, protocolVersion)
+        else null
+
+      if (serializedValue != null)
+        bytesCount += serializedValue.remaining()
```

Comment:
Unfortunately this is the critical path. We shouldn't slow it down just to make the code look pretty :smile: